dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,859 @@
+                if (SUCCEEDED(hr) && !Background)
+                {
+                    --pos;
+                    InsertMenuItem(qcmi, pos, id, 0, -1); // Separator
+                }
+            }
+            if (SUCCEEDED(hr))
+            {
+                qcmi.idCmdFirst = id + 1;
+                hr = S_FALSE; // Don't add verbs
+            }
+            break;
+        }
+
+        case DFM_INVOKECOMMAND:
+        {
+            hr = S_FALSE;
+            CCabFolder *pCabFolder = static_cast<CCabFolder*>(psf);
+            switch (wParam)
+            {
+                case IDC_EXTRACT:
+                case IDC_EXTRACTALL:
+                    hr = pCabFolder->ExtractFilesUI(hwnd, wParam == IDC_EXTRACT ? pdtobj : NULL);
+                    break;
+            }
+            break;
+        }
+    }
+    return hr;
+}
+
+static HRESULT CALLBACK FolderBackgroundMenuCallback(IShellFolder *psf, HWND hwnd,
+                                                     IDataObject *pdtobj, UINT uMsg,
+                                                     WPARAM wParam, LPARAM lParam)
+{
+    return ItemMenuCallback(psf, hwnd, NULL, uMsg, wParam, lParam);
+}
+
+int CEnumIDList::FindNamedItem(PCUITEMID_CHILD pidl) const
+{
+    CABITEM *needle = (CABITEM*)pidl;
+    for (ULONG i = 0, c = GetCount(); i < c; ++i)
+    {
+        CABITEM *item = (CABITEM*)DPA_FastGetPtr(m_Items, i);
+        if (!lstrcmpiW(needle->Path, item->Path))
+            return i;
+    }
+    return -1;
+}
+
+struct FILLCALLBACKDATA
+{
+    CEnumIDList *pEIDL;
+    SHCONTF ContF;
+};
+
+static HRESULT CALLBACK EnumFillCallback(EXTRACTCALLBACKMSG msg, const EXTRACTCALLBACKDATA &ecd, LPVOID cookie)
+{
+    FILLCALLBACKDATA &data = *(FILLCALLBACKDATA*)cookie;
+
+    switch ((UINT)msg)
+    {
+        case ECM_FILE:
+        {
+            const FDINOTIFICATION &fdin = *ecd.pfdin;
+            HRESULT hr = S_FALSE;
+            SFGAOF attr = MapFSToSFAttributes(fdin.attribs & CABITEM::FSATTS);
+            if (IncludeInEnumIDList(data.ContF, attr))
+            {
+                UINT datetime = MAKELONG(fdin.time, fdin.date);
+                CABITEM *item = CreateItem(fdin.psz1, fdin.attribs, fdin.cb, datetime);
+                if (!item)
+                    return E_OUTOFMEMORY;
+                if (FAILED(hr = data.pEIDL->Append((LPCITEMIDLIST)item)))
+                    SHFree(item);
+            }
+            return SUCCEEDED(hr) ? S_FALSE : hr; // Never extract
+        }
+    }
+    return E_NOTIMPL;
+}
+
+HRESULT CEnumIDList::Fill(LPCWSTR path, HWND hwnd, SHCONTF contf)
+{
+    FILLCALLBACKDATA data = { this, contf };
+    return ExtractCabinet(path, NULL, EnumFillCallback, &data);
+}
+
+HRESULT CEnumIDList::Fill(PCIDLIST_ABSOLUTE pidl, HWND hwnd, SHCONTF contf)
+{
+    WCHAR path[MAX_PATH];
+    if (SHGetPathFromIDListW(pidl, path))
+        return Fill(path, hwnd, contf);
+    return HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);
```

Comment:
Do these SHGetPathFromIDListW and similar set the last error to something meaningful when they fail, or not? (Granted, MSDN doesn't seem to mention anything in this regard...)

Reply:
No, this design is from Win95. You just have to invent your own HRESULT that makes sense.